required.go

```diff
@@ -48,49 +49,87 @@ func NewRequiredValidator(definition RequiredValidatorDefinition) (RequiredValid
 	return RequiredValidator{definition}, nil
 }
 
+// Validate reports whether input is valid against required keys.
+// Fields of Input struct must be exported.
 func (r RequiredValidator) Validate(input interface{}) error {
-	var v reflect.Value
-	if reflect.TypeOf(input).Kind() != reflect.Ptr {
-		v = reflect.ValueOf(input)
-	} else {
-		v = reflect.ValueOf(input).Elem()
+	v, ok := convertToConcreteValue(reflect.ValueOf(input))
+	if !ok {
+		return &InvalidTypeError{
+			Definition: r.definition,
+			Input:      input,
+		}
 	}
 	if v.Kind() != reflect.Struct {
-		return &InvalidFieldTypeError{
+		return &InvalidTypeError{
 			Definition: r.definition,
 			Input:      input,
 		}
 	}
-
 	for _, key := range r.definition.Required {
-		e := v.FieldByName(key)
-		if !e.IsValid() {
-			return &InvalidFieldTypeError{
-				Definition: r.definition,
-				Input:      input,
-			}
-		}
-		n, ok := e.Interface().(driver.Valuer)
+		e, ok := getFieldByName(v, key)
 		if !ok {
 			return &InvalidFieldTypeError{
 				Definition: r.definition,
 				Input:      input,
 			}
 		}
-		v, err := n.Value()
-		if err != nil {
+		c, ok := convertToConcreteValue(e)
+		if !ok {
 			return &InvalidFieldTypeError{
 				Definition: r.definition,
 				Input:      input,
 			}
 		}
-		if v == nil {
+		i := c.Interface()
+		ok = isValid(i)
+		if !ok {
 			return &RequiredValidationError{
 				Definition: r.definition,
 				Input:      input,
 			}
 		}
 	}
-
 	return nil
 }
+
+// convertToConcreteValue returns a concrete value that stored in the pointer.
+// The ok return value reports whether conversion was successful.
+func convertToConcreteValue(input reflect.Value) (value reflect.Value, ok bool) {
+	if input.Kind() != reflect.Ptr {
+		return input, true
+	}
+	if input.IsNil() {
+		return reflect.Value{}, false
+	}
+	return input.Elem(), true
+}
+
+// getFieldByName returns the struct field with the given name.
+// The ok return value reports whether field with key was found in value.
+func getFieldByName(v reflect.Value, key string) (f reflect.Value, ok bool) {
```

Comment:
Replace `v` to `s`, if `v` means a instance of struct.
Replace `f` to `v`, if `f` means a value of the field provided with `key` in struct.